dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/receiver/ThreadSafeReceiver.kt

```diff
@@ -0,0 +1,35 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.receiver
+
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import java.util.concurrent.atomic.AtomicBoolean
+
+internal abstract class ThreadSafeReceiver : BroadcastReceiver() {
+
+    val isRegistered = AtomicBoolean(false)
+
+    fun registerForFilter(
+        context: Context,
+        filter: IntentFilter,
+        doOnRegister: ((intent: Intent?) -> Unit)? = null
+    ) {
+
+        val intent = context.registerReceiver(this, filter)
+        doOnRegister?.invoke(intent)
+        isRegistered.set(true)
+    }
+
+    fun checkAndUnregister(context: Context) {
+        if (isRegistered.compareAndSet(true, false)) {
+            context.unregisterReceiver(this)
+        }
+    }
```

Comment:
again the `check` part in the name is not necessary, you can simply call it `unregisterReceiver`